### PR TITLE
Refactor driver selection

### PR DESCRIPTION
**Pull request by imurchie**
## Proposed changes

On a boring flight with broken WiFi, I got to thinking about why the driver configuration is in the code at all. Couldn't it be in a configuration file like a sane system?

This is a first iteration on making it so. It is just what popped into my head, and I'm sure there are many improvements.

## Types of changes

What types of changes does your code introduce to Appium?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have read the [CONTRIBUTING](https://github.com/appium/appium/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla.js.foundation/appium/appium)
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Review by jlipps — COMMENTED** [review]
I love this idea. I have not yet reviewed the code in detail, but what strikes me initially is that this a step on the way to Appium 2.0. I wonder if it would be easy enough to add a few more bits to this PR to basically get us what we need for "driver independence". IIRC, this is what else we'd need:

1. Ability to override location of drivers.json through a server flag (so somebody could start up appium with their own drivers.json, or so that Appium 2.0 could decide that this file belongs in `~/.appium-drivers.json` or similar
2. Ability to write new driver metadata into driver.json (so that in the future running `appium driver install foo` will add the driver foo
3. Ability to designate which drivers are installed vs just available to install. We could implement this a few different ways, for example having an 'installed' flag in drivers.json, or having essentially two files: `available-drivers.json` (which has metadata for the official appium drivers), and `drivers.json` which has metadata for only the installed drivers.

I understand these features would significantly increase the scope of this change, so I'm fine if we want to just review this as is and add the other stuff in a bit. But it struck me that this is already pretty close and at least wanted to make sure we have an eye to the future when reviewing the specifics of this proposal.

File: drivers.json [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+  "AndroidDriver": {
+    "driverClass": "AndroidDriver",
+    "packageName": "appium-android-driver",
+    "platformName": "Android",
+    "repository": "https://github.com/appium/appium-android-driver",
+    "constraints": [{
+      "type": "capability",
+      "capability": "platformVersion",
+      "condition": ">=6.0.0",
+      "alternativeDriver": "AndroidDriver",
```
Comment (mykola-mokhnach) [inline drivers.json]: Perhaps `AndroidUiautomator2Driver` ?

Comment (mykola-mokhnach) [reply drivers.json]: Also, it should be possible to set multiple alernatives

File: drivers.json [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    "repository": "https://github.com/appium/appium-fake-driver"
+  },
+  "AndroidDriver": {
+    "driverClass": "AndroidDriver",
+    "packageName": "appium-android-driver",
+    "platformName": "Android",
+    "repository": "https://github.com/appium/appium-android-driver",
+    "constraints": [{
+      "type": "capability",
+      "capability": "platformVersion",
```
Comment (mykola-mokhnach) [inline drivers.json]: name?

File: drivers.json [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+  },
+  "IosDriver": {
+    "driverClass": "IosDriver",
+    "packageName": "appium-ios-driver",
+    "platformName": "iOS",
+    "repository": "https://github.com/appium/appium-ios-driver",
+    "constraints": [{
+      "type": "capability",
+      "capability": "platformVersion",
+      "condition": ">=10.0.0",
```
Comment (mykola-mokhnach) [inline drivers.json]: should the switch to the alternative driver be enforced or just a warning message should be printed?

File: drivers.json [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    "driverClass": "XCUITestDriver",
+    "packageName": "appium-xcuitest-driver",
+    "automationName": "XCUITest",
+    "repository": "https://github.com/appium/appium-xcuitest-driver"
+  },
+  "YouiEngineDriver": {
+    "driverClass": "YouiEngineDriver",
+    "packageName": "appium-youiengine-driver",
+    "automationName": "YouiEngine",
+    "repository": ""
```
Comment (mykola-mokhnach) [inline drivers.json]: I'd rather just keep it unset

File: lib/driver-utils.js [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+  } catch (err) {
+    log.errorAndThrow(`Unable to load driver '${driverClass}' from package '${packageName}': ${err.message}`);
+  }
+};
+
+function getDriverFromConstraints (caps, driverInfo) {
+  const getAlternativeDriver = function getAlternativeDriver (driverInfo) {
+    let alternativeDriver;
+    for (const constraint of (driverInfo.constraints || [])) {
+      if (constraint.type === 'capability') {
```
Comment (mykola-mokhnach) [inline lib/driver-utils.js]: `switch` might be more readable

File: lib/driver-utils.js [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+        if (constraint.capability === 'platformVersion') {
+          const platformVersion = semver.valid(semver.coerce(caps.platformVersion));
```
Comment (mykola-mokhnach) [inline lib/driver-utils.js]: is it by design that `caps` variable is caught in the closure? Shall we use _.memoize instead?

File: lib/driver-utils.js [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+              log.warn(constraint.message);
+            }
+            if (constraint.alternativeDriver) {
+              alternativeDriver = constraint.alternativeDriver;
+            }
+          }
+        }
+      }
+    }
+    return alternativeDriver;
```
Comment (mykola-mokhnach) [inline lib/driver-utils.js]: I cannot understand this logic

File: lib/driver-utils.js [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+    const modifiedPackageName = DRIVER_MAP[alternativeDriver].packageName;
+    if (modifiedDriverClass && modifiedPackageName) {
+      return [modifiedPackageName, modifiedDriverClass];
+    }
+  }
+  return [];
+}
+
+function getDriverFromAutomationName (caps) {
+  let packageName;
```
Comment (mykola-mokhnach) [inline lib/driver-utils.js]: if (!_.isString(caps.automationName)) return [null, null];

File: lib/driver-utils.js [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+      packageName = info.packageName;
+
+      // now check if there are constraints on this driver
+      const [modifiedPackageName, modifiedDriverClass] = getDriverFromConstraints(caps, info);
+      if (modifiedPackageName && modifiedDriverClass) {
+        packageName = modifiedPackageName;
+        driverClass = modifiedDriverClass;
+      }
+      // no need to go further
+      break;
```
Comment (mykola-mokhnach) [inline lib/driver-utils.js]: we can return right here and get rid of intermediate variables

File: lib/driver-utils.js [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+    packageName = info.packageName;
+
+    // now check if there are constraints on this driver
+    const [modifiedPackageName, modifiedDriverClass] = getDriverFromConstraints(caps, info);
+    if (modifiedPackageName && modifiedDriverClass) {
+      packageName = modifiedPackageName;
+      driverClass = modifiedDriverClass;
+    }
+    // no need to go further
+    break;
```
Comment (mykola-mokhnach) [inline lib/driver-utils.js]: we could return right here

**Comment by jlipps** [comment]
@imurchie let me know about my high-level comments and then I'm happy to do an in-depth review as well.

File: lib/driver-utils.js (hunk 9 above)
Comment (farhanmannan) [reply lib/driver-utils.js]: nice

**Comment by mykola-mokhnach** [comment]
@imurchie Do you plan to finish this PR?

**Comment by imurchie** [comment]
Yes. Just haven't had time.

**Pull request closed without merging**